File: Patrones/Estructurales/DAO/Java/src/UsuarioDAOImpl.java
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class UsuarioDAOImpl implements UsuarioDAO {
    private Connection conexion;

    public UsuarioDAOImpl(Connection conexion) {
        this.conexion = conexion;
    }

    @Override
    public Usuario obtenerUsuarioPorId(int id) {
        Usuario usuario = null;
        String query = "SELECT * FROM usuarios WHERE id = ?";
        try (PreparedStatement statement = conexion.prepareStatement(query)) {
            statement.setInt(1, id);
            ResultSet rs = statement.executeQuery();
            if (rs.next()) {
                usuario = new Usuario(rs.getInt("id"), rs.getString("nombre"), rs.getString("email"));
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return usuario;
    }

    @Override
    public List<Usuario> obtenerTodosLosUsuarios() {
        List<Usuario> usuarios = new ArrayList<>();
        String query = "SELECT * FROM usuarios";
        try (Statement statement = conexion.createStatement()) {
            ResultSet rs = statement.executeQuery(query);
            while (rs.next()) {
                Usuario usuario = new Usuario(rs.getInt("id"), rs.getString("nombre"), rs.getString("email"));
                usuarios.add(usuario);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return usuarios;
    }

    @Override
    public void insertarUsuario(Usuario usuario) {
        String query = "INSERT INTO usuarios (nombre, email) VALUES (?, ?)";
        try (PreparedStatement statement = conexion.prepareStatement(query)) {
            statement.setString(1, usuario.getNombre());
            statement.setString(2, usuario.getEmail());
            statement.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void actualizarUsuario(Usuario usuario) {
        String query = "UPDATE usuarios SET nombre = ?, email = ? WHERE id = ?";
        try (PreparedStatement statement = conexion.prepareStatement(query)) {
            statement.setString(1, usuario.getNombre());
            statement.setString(2, usuario.getEmail());
            statement.setInt(3, usuario.getId());
            statement.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void eliminarUsuario(int id) {
        String query = "DELETE FROM usuarios WHERE id = ?";
        try (PreparedStatement statement = conexion.prepareStatement(query)) {
            statement.setInt(1, id);
            statement.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: Patrones/Estructurales/DAO/Python/UsuarioDAOImpl.py
import mysql.connector
from mysql.connector import Error
from Usuario import Usuario

class UsuarioDAOImpl:
    def __init__(self, conexion):
        self.__conexion = conexion

    def obtener_usuario_por_id(self, id):
        usuario = None
        query = "SELECT * FROM usuarios WHERE id = %s"
        try:
            cursor = self.__conexion.cursor()
            cursor.execute(query, (id,))
            row = cursor.fetchone()
            if row:
                usuario = Usuario(row[0], row[1], row[2])
        except Error as e:
            print(f"Error al obtener usuario por ID: {e}")
        return usuario

    def obtener_todos_los_usuarios(self):
        usuarios = []
        query = "SELECT * FROM usuarios"
        try:
            cursor = self.__conexion.cursor()
            cursor.execute(query)
            rows = cursor.fetchall()
            for row in rows:
                usuario = Usuario(row[0], row[1], row[2])
                usuarios.append(usuario)
        except Error as e:
            print(f"Error al obtener todos los usuarios: {e}")
        return usuarios

    def insertar_usuario(self, usuario):
        query = "INSERT INTO usuarios (nombre, email) VALUES (%s, %s)"
        try:
            cursor = self.__conexion.cursor()
            cursor.execute(query, (usuario.nombre, usuario.email))
            self.__conexion.commit()
        except Error as e:
            print(f"Error al insertar usuario: {e}")

    def actualizar_usuario(self, usuario):
        query = "UPDATE usuarios SET nombre = %s, email = %s WHERE id = %s"
        try:
            cursor = self.__conexion.cursor()
            cursor.execute(query, (usuario.nombre, usuario.email, usuario.id_usuario))
            self.__conexion.commit()
        except Error as e:
            print(f"Error al actualizar usuario: {e}")

    def eliminar_usuario(self, id):
        query = "DELETE FROM usuarios WHERE id = %s"
        try:
            cursor = self.__conexion.cursor()
            cursor.execute(query, (id,))
            self.__conexion.commit()
        except Error as e:
            print(f"Error al eliminar usuario: {e}")
